packages/app/lib/src/features/home/application/feed_service.dart

```diff
@@ -18,70 +17,86 @@ part 'feed_service.g.dart';
-      return updatedCachedPost;
-    }
+  /// Returns `true` if more were fetched, and false if not.
+  Future<bool> fetchMore() async {
+    final postRepo = ref.read(postRepositoryProvider);
+
+    // Fetch the specific batch
+    final fetchedPosts = await postRepo.readPosts(feed, state.cursorPos);
 
-    // Get user's location.
-    final posts = await postRepo.readPosts(state.cursorPos);
+    if (fetchedPosts.isEmpty) return false;
 
-    if (posts.isEmpty) return null;
+    // Store each fetched post in WipPost and collect their IDs
+    final newPostIds = <PostId>[];
+    for (final post in fetchedPosts) {
+      // Store the post in WipPost
+      ref.read(wipPostProvider(post.id).notifier).setPost(post);
 
-    // Gets post ID from first post
-    final postId = posts[postIndex].id;
+      // Collect the post ID
+      newPostIds.add(post.id);
+    }
 
-    // Adds posts and ids to their respective list
+    // Update the state with the new batch of post IDs and cursor
     state = state.copyWith(
-      posts: [...state.posts, ...posts.map((tuple) => tuple.entity)],
-      ids: [...state.ids, ...posts.map((tuple) => tuple.id)],
-      cursorPos: posts.lastOrNull?.id,
+      ids: [...state.ids, ...newPostIds],
+      cursorPos: newPostIds.lastOrNull,
     );
-    final post = state.posts.elementAtOrNull(postIndex);
 
-    // Adds post id to post
-    final updatedPost = post?.copyWith(id: postId);
-    return updatedPost ?? await fetch(postIndex + 1);
+    return true;
   }
 }
 
-/// Fetch a single post from the feed.
+/// Fetches a single post from the feed.
 ///
-/// [postIndex] is the index of the current post user is viewing.
+/// [postIndex] is the index of the current post the user is viewing.
 @riverpod
-FutureOr<PostEntity?> singlePost(
+FutureOr<PostId?> singlePost(
   Ref ref,
   FeedEntity feed,
   int postIndex,
 ) async {
   // Keep previous posts in cache to make scrolling up possible.
-  // Otherwise, `ListView` freaks out.
+  // Otherwise, the `ListView` freaks out.
   if (postIndex != 0) {
-    final post = await ref.watch(
-      singlePostProvider(feed, postIndex - 1).future,
+    await ref.watch(
+      singlePostProvider(feed, postIndex - 1).selectAsync((_) {}),
     );
-    // Keeps avatar in local memory
-    ref.watch(avatarServiceProvider(post?.authorName));
   }
 
-  final feedNotifier = ref.watch(feedServiceProvider(feed).notifier);
-  final post = await feedNotifier.fetch(postIndex);
+  final ids = ref.watch(feedServiceProvider(feed).select((s) => s.ids));
+
+  var next = ids.elementAtOrNull(postIndex);
+  var moreToGet = true;
 
-  return post;
+  while (moreToGet && next == null) {
+    moreToGet = await ref.watch(feedServiceProvider(feed).notifier).fetchMore();
+    next = ids.elementAtOrNull(postIndex);
+  }
+
+  return next;
+}
+
+/// Store a post independently of any feed for memory efficiency.
+///
+/// [id] is the unique identifier for the post.
+@Riverpod(keepAlive: true)
```

Comment:
I'd like to figure out how to do this without needing to use keepAlive, so that they clean themselves up when not needed.